File: src/main/java/br/com/dusty/dkits/listener/quit/PlayerQuitListener.java
package br.com.dusty.dkits.listener.quit;

import br.com.dusty.dkits.gamer.EnumRank;
import br.com.dusty.dkits.gamer.Gamer;
import br.com.dusty.dkits.gamer.GamerRegistry;
import br.com.dusty.dkits.util.ScoreboardUtils;
import br.com.dusty.dkits.util.TaskUtils;
import br.com.dusty.dkits.util.text.Text;
import br.com.dusty.dkits.util.web.WebAPI;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerQuitEvent;

public class PlayerQuitListener implements Listener {
	
	private static final String QUIT_MESSAGE_PREFIX = Text.neutralOf("[").negative("-").neutral("] ").toString();
	
	@EventHandler
	public void onPlayerQuit(PlayerQuitEvent event) {
		Player player = event.getPlayer();
		
		Gamer gamer = GamerRegistry.unregister(player);
		
		if(gamer.isCombatTagged()){
			Gamer combatPartner = gamer.getCombatPartner();
			
			if(combatPartner != null){
				combatPartner.addKill();
				combatPartner.addKillMoney();
				combatPartner.addKillXp();
				
				gamer.addDeath();
				gamer.removeDeathMoney();
				gamer.removeDeathXp();
			}
			
			Bukkit.broadcastMessage(Text.negativeOf(player.getName())
			                            .neutral(" deslogou em ")
			                            .negative("combate")
			                            .neutral("!")
			                            .toString());
		}
		
		TaskUtils.async(() -> WebAPI.saveProfiles(gamer));
		
		if(gamer.getRank().isLowerThan(EnumRank.MOD))
			event.setQuitMessage(QUIT_MESSAGE_PREFIX + player.getName());
		else
			event.setQuitMessage(null);
		
		ScoreboardUtils.updateAll();
	}
}

File: src/main/kotlin/br/com/dusty/dkits/listener/quit/PlayerQuitListener.kt
package br.com.dusty.dkits.listener.quit

import br.com.dusty.dkits.Config
import br.com.dusty.dkits.clan.ClanRegistry
import br.com.dusty.dkits.gamer.EnumRank
import br.com.dusty.dkits.gamer.GamerRegistry
import br.com.dusty.dkits.scoreboard.Scoreboards
import br.com.dusty.dkits.util.Tasks
import br.com.dusty.dkits.util.entity.gamer
import br.com.dusty.dkits.util.stdlib.clearFormatting
import br.com.dusty.dkits.util.text.Text
import br.com.dusty.dkits.util.web.WebAPI
import br.com.dusty.dkits.warp.Warps
import org.bukkit.Bukkit
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerQuitEvent

object PlayerQuitListener: Listener {

	private val QUIT_MESSAGE_PREFIX = Text.basicOf("[").negative("-").basic("] ").toString()

	@EventHandler
	fun onPlayerQuit(event: PlayerQuitEvent) {
		val player = event.player
		val gamer = player.gamer()

		GamerRegistry.unregister(player.uniqueId)?.run {
			if (isCombatTagged()) {
				Bukkit.broadcastMessage(Text.negativePrefix().negative(player.displayName.clearFormatting()).basic(" deslogou em ").negative("combate").basic("!").toString())

				combatPartner?.kill(this)
			}

			warp.dispatchGamer(this, Warps.NONE)

			if (Config.data.serverStatus == Config.EnumServerStatus.ONLINE) Tasks.async(Runnable { WebAPI.saveProfiles(gamer) })

			if (clan != null) {
				val clan = clan!!

				clan.onlineMembers.remove(this)

				if (clan.leader == this) clan.leader = null

				if (clan.onlineMembers.isEmpty()) ClanRegistry.CLAN_BY_STRING.remove(clan.uuid)
			}

			if (rank.isLowerThan(EnumRank.MOD)) event.quitMessage = QUIT_MESSAGE_PREFIX + player.displayName.clearFormatting()
			else event.quitMessage = null
		}

		Scoreboards.update()
	}
}
